### ISSUE TITLE
Passing a non-`const` trait function item value to a `const fn` is no longer allowed in a constant context

### ISSUE THREAD

**Issue by yvt** [rust-lang/rust#104155]
### Code

I tried this code:

```rust
const _: () = core::mem::forget(Box::<u32>::default);       // ERROR
const _: () = core::mem::forget(|| Box::<u32>::default());  // ok
```

([Playground](https://play.rust-lang.org/?version=beta&mode=debug&edition=2021&gist=f846e2e1b06b1d52476801d0a98874f2))

I expected to see this happen: Successful compilation

Instead, this happened:

```
error[E0277]: the trait bound `Box<u32>: Default` is not satisfied
 --> src/lib.rs:1:15
  |
1 | const _: () = core::mem::forget(Box::<u32>::default);
  |               ^^^^^^^^^^^^^^^^^ the trait `~const Default` is not implemented for `Box<u32>`
  |
note: the trait `Default` is implemented for `Box<u32>`, but that implementation is not `const`
 --> src/lib.rs:1:15
  |
1 | const _: () = core::mem::forget(Box::<u32>::default);
  |               ^^^^^^^^^^^^^^^^^
```

### Version it worked on

It most recently worked on: 1.65.0

### Version with regression

`rustc --version --verbose`:
```
rustc 1.66.0-beta.1 (e080cc5a6 2022-11-01)
binary: rustc
commit-hash: e080cc5a659fb760c0bc561b722a790dad35b5e1
commit-date: 2022-11-01
host: aarch64-apple-darwin
release: 1.66.0-beta.1
LLVM version: 15.0.2
```

@rustbot modify labels: +regression-from-stable-to-beta -regression-untriaged

**Comment by jruderman** [comment]
This became an error due to #100982 (@fee1-dead, @oli-obk).

<details><summary>Bisect info</summary>

searched nightlies: from nightly-2022-09-18 to nightly-2022-11-07
regressed nightly: nightly-2022-09-23
searched commit range: https://github.com/rust-lang/rust/compare/9062b780b32d2eab060b4432863e085d9504ca5c...e7119a0300b87a3d670408ee8e847c6821b3ae80
regressed commit: https://github.com/rust-lang/rust/commit/7a8636c843bd24038fe1d1f69b4a8e4b0ea55d4e

bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.4

Host triple: `x86_64-apple-darwin`

I used `cargo bisect-rustc --regress error` on the following file

```rust
const _: () = core::mem::forget(Box::<u32>::default); 
fn main(){}
```

</details>

**Comment by inquisitivecrystal** [comment]
Assigning priority as [discussed](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23104155.20Passing.20a.20non-.60const.60.20trait.20function.20item.20value.20to.20.E2.80.A6) by the Prioritization Working Group.

**Comment by yvt** [comment]
Can this be closed now that beta backport (#104447) is complete?
